src/aspire/classification/align2d.py

```diff
@@ -225,35 +376,620 @@ def align(self, classes, reflections, basis_coefficients):
+        """
+        Performs the actual rotational alignment estimation,
+        returning parameters needed for averaging.
+        """
+
+        # Admit simple case of single case alignment
+        classes = np.atleast_2d(classes)
+        reflections = np.atleast_2d(reflections)
+
+        n_classes = classes.shape[0]
+
+        # Instantiate matrices for results
+        rotations = np.zeros(classes.shape, dtype=self.dtype)
+        correlations = np.zeros(classes.shape, dtype=self.dtype)
+        shifts = np.zeros((*classes.shape, 2), dtype=int)
+
+        for k in trange(n_classes):
+            # # Get the array of images for this class, using the `alignment_src`.
+            images = self._cls_images(classes[k], src=self.alignment_src)
+
+            self._reddychatterji(
+                k, images, classes, reflections, rotations, correlations, shifts
+            )
+
+        return classes, reflections, rotations, shifts, correlations
+
+    def _reddychatterji(
+        self, k, images, classes, reflections, rotations, correlations, shifts
+    ):
+        """
+        Compute the Reddy Chatterji method registering images[1:] to image[0].
+
+        This differs from papers and published scikit implimentations by
+        computing the fixed base image[0] pipeline once then reusing.
+        """
+
+        # De-Mean
+        images -= images.mean(axis=(-1, -2))[:, np.newaxis, np.newaxis]
+
+        # Precompute fixed_img data used repeatedly in the loop below.
+        fixed_img = images[0]
+        # Difference of Gaussians (Band Filter)
+        fixed_img_dog = difference_of_gaussians(fixed_img, 1, 4)
+        # Window Images (Fix spectral boundary)
+        wfixed_img = fixed_img_dog * window("hann", fixed_img.shape)
+        # Transform image to Fourier space
+        fixed_img_fs = np.abs(np.fft.fftshift(np.fft.fft2(wfixed_img))) ** 2
+        # Compute Log Polar Transform
+        radius = fixed_img_fs.shape[0] // 8  # Low Pass
+        warped_fixed_img_fs = warp_polar(
+            fixed_img_fs,
+            radius=radius,
+            output_shape=fixed_img_fs.shape,
+            scaling="log",
+        )
+        # Only use half of FFT, because it's symmetrical
+        warped_fixed_img_fs = warped_fixed_img_fs[: fixed_img_fs.shape[0] // 2, :]
+
+        # Now prepare for rotating original images,
+        #   and searching for translations.
+        # We start back at the raw fixed_img.
+        twfixed_img = fixed_img * window("hann", fixed_img.shape)
+
+        # Register image `m` against image[0]
+        for m in range(1, len(images)):
+            # Get the image to register
+            regis_img = images[m]
+
+            # Reflect images when necessary
+            if reflections[k][m]:
+                regis_img = np.flipud(regis_img)
+
+            # Difference of Gaussians (Band Filter)
+            regis_img_dog = difference_of_gaussians(regis_img, 1, 4)
```

Comment:
we need all the intermediaries. It would be a mess. (i can stuff them in a class or a dict, but you are kidding yourself if you think that will be less/clearer code).

Reply:
Ok got it.